File: trees/99_RecoverBST.cc
#include <vector>
#include <iostream>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

//O(n) space solution
class Solution {
public:
    vector<TreeNode*> nodes;
    void recoverTree(TreeNode* root) {
        this->inOrderTraverse(root);
        int dip1 = -1, dip2 = -1;
        for(int i = 0; i < this->nodes.size(); i++){
            if(i + 1 < this->nodes.size() && this->nodes[i+1]->val < this->nodes[i]->val){
                if(dip1 != -1){
                    dip2 = i+1;
                }
                else{
                    dip1 = i;
                }
            }
        }
        if(dip2 != -1){
            int val1 = nodes[dip1]->val,val2 = nodes[dip2]->val;
            nodes[dip1]->val = val2;
            nodes[dip2]->val = val1;
        }
        else{
            int val1 = nodes[dip1]->val,val2 = nodes[dip1+1]->val;
            nodes[dip1]->val = val2;
            nodes[dip1+1]->val = val1;
        }
    }

    void inOrderTraverse(TreeNode* node){
        if(node && node->left) this->inOrderTraverse(node->left);
        if(node) this->nodes.push_back(node);
        if(node && node->right) this->inOrderTraverse(node->right);
    }
};

File: trees/99_RecoverBST.py
class Solution:
    """O(1) Space with Morris Inorder Predecessor Traversal"""
    def recoverTree(self, root):
        # predecessor is a Morris predecessor. 
        # In the 'loop' cases it could be equal to the node itself predecessor == root.
        # pred is a 'true' predecessor, 
        # the previous node in the inorder traversal.
        x = y = predecessor = pred = None
        
        while root:
            # If there is a left child
            # then compute the predecessor.
            # If there is no link predecessor.right = root --> set it.
            # If there is a link predecessor.right = root --> break it.
            if root.left:       
                # Predecessor node is one step left 
                # and then right till you can.
                predecessor = root.left
                while predecessor.right and predecessor.right != root:
                    predecessor = predecessor.right
 
                # set link predecessor.right = root
                # and go to explore left subtree
                if predecessor.right is None:
                    predecessor.right = root
                    root = root.left
                # break link predecessor.right = root
                # link is broken : time to change subtree and go right
                else:
                    # check for the swapped nodes
                    if pred and root.val < pred.val:
                        y = root
                        if x is None:
                            x = pred 
                    pred = root
                    
                    predecessor.right = None
                    root = root.right
            # If there is no left child
            # then just go right.
            else:
                # check for the swapped nodes
                if pred and root.val < pred.val:
                    y = root
                    if x is None:
                        x = pred 
                pred = root
                
                root = root.right
        
        x.val, y.val = y.val, x.val